File: python/tests/test_formatter.py
"""Tests for JSON formatter."""

import json
import pytest
from datetime import datetime, timezone
from agora_log import initialize, get_logger, LogConfig


class TestJSONFormatterOutput:
    """Tests for JSON output format."""

    def test_output_is_valid_json(self, console_config: LogConfig, capsys):
        """Test that output is valid JSON."""
        logger = initialize(console_config)

        logger.info("test message")

        captured = capsys.readouterr()
        if captured.out:
            # Should be parseable as JSON
            entry = json.loads(captured.out.strip())

            assert entry is not None
            assert isinstance(entry, dict)

    def test_json_one_entry_per_line(self, console_config: LogConfig, capsys):
        """Test that each log entry is on its own line."""
        logger = initialize(console_config)

        logger.info("message 1")
        logger.info("message 2")
        logger.info("message 3")

        captured = capsys.readouterr()
        if captured.out:
            lines = [line for line in captured.out.strip().split("\n") if line]

            assert len(lines) == 3

            # Each line should be valid JSON
            for line in lines:
                entry = json.loads(line)
                assert "message" in entry


class TestRequiredFields:
    """Tests for required fields in JSON output."""

    def test_has_timestamp_field(self, console_config: LogConfig, capsys):
        """Test that timestamp field is present."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "timestamp" in entry
            assert entry["timestamp"] is not None

    def test_has_level_field(self, console_config: LogConfig, capsys):
        """Test that level field is present."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "level" in entry
            assert entry["level"] == "INFO"

    def test_has_message_field(self, console_config: LogConfig, capsys):
        """Test that message field is present."""
        logger = initialize(console_config)

        logger.info("test message")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "message" in entry
            assert entry["message"] == "test message"

    def test_has_service_field(self, console_config: LogConfig, capsys):
        """Test that service field is present."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "service" in entry
            assert entry["service"] == "test-service"

    def test_has_file_field(self, console_config: LogConfig, capsys):
        """Test that file field is present (REQUIRED)."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "file" in entry
            assert entry["file"] is not None
            assert len(entry["file"]) > 0

    def test_has_line_field(self, console_config: LogConfig, capsys):
        """Test that line field is present (REQUIRED)."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "line" in entry
            assert isinstance(entry["line"], int)
            assert entry["line"] > 0

    def test_has_function_field(self, console_config: LogConfig, capsys):
        """Test that function field is present (REQUIRED)."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "function" in entry
            assert entry["function"] is not None
            assert len(entry["function"]) > 0


class TestOptionalFields:
    """Tests for optional fields in JSON output."""

    def test_environment_field(self, console_config: LogConfig, capsys):
        """Test that environment field is included."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "environment" in entry
            assert entry["environment"] == "development"

    def test_version_field(self, console_config: LogConfig, capsys):
        """Test that version field is included."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "version" in entry
            assert entry["version"] == "1.0.0"

    def test_host_field(self, console_config: LogConfig, capsys):
        """Test that host field is included."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "host" in entry
            assert len(entry["host"]) > 0

    def test_logger_name_field(self, console_config: LogConfig, capsys):
        """Test that logger_name field is included."""
        logger = initialize(console_config)

        named_logger = get_logger("test.module.component")
        named_logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "logger_name" in entry
            assert entry["logger_name"] == "test.module.component"


class TestTimestampFormat:
    """Tests for timestamp format."""

    def test_timestamp_is_iso8601(self, console_config: LogConfig, capsys):
        """Test that timestamp is in ISO 8601 format."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            timestamp = entry["timestamp"]

            # Should contain 'T' separator
            assert "T" in timestamp

            # Should be parseable as ISO 8601
            dt = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
            assert dt is not None

    def test_timestamp_has_timezone(self, console_config: LogConfig, capsys):
        """Test that timestamp includes timezone."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            timestamp = entry["timestamp"]

            # Should end with Z or contain timezone offset
            assert timestamp.endswith("Z") or "+" in timestamp or timestamp.endswith(
                "+00:00"
            )

    def test_timestamp_has_microseconds(self, console_config: LogConfig, capsys):
        """Test that timestamp includes microseconds."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            timestamp = entry["timestamp"]

            # ISO 8601 with microseconds has format: YYYY-MM-DDTHH:MM:SS.ffffff
            # Should contain decimal point for fractional seconds
            assert "." in timestamp or "," in timestamp


class TestContextSerialization:
    """Tests for context field serialization."""

    def test_context_is_dict(self, console_config: LogConfig, capsys):
        """Test that context field is a dictionary."""
        logger = initialize(console_config)

        logger.info("test", key="value")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "context" in entry
            assert isinstance(entry["context"], dict)

    def test_context_serializes_strings(self, console_config: LogConfig, capsys):
        """Test that context serializes string values."""
        logger = initialize(console_config)

        logger.info("test", name="John", role="admin")

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert entry["context"]["name"] == "John"
            assert entry["context"]["role"] == "admin"

    def test_context_serializes_numbers(self, console_config: LogConfig, capsys):
        """Test that context serializes numeric values."""
        logger = initialize(console_config)

        logger.info("test", count=42, price=19.99)

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert entry["context"]["count"] == 42
            assert entry["context"]["price"] == 19.99

    def test_context_serializes_booleans(self, console_config: LogConfig, capsys):
        """Test that context serializes boolean values."""
        logger = initialize(console_config)

        logger.info("test", active=True, deleted=False)

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert entry["context"]["active"] is True
            assert entry["context"]["deleted"] is False

    def test_context_serializes_null(self, console_config: LogConfig, capsys):
        """Test that context serializes None as null."""
        logger = initialize(console_config)

        logger.info("test", optional=None)

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "optional" in entry["context"]
            assert entry["context"]["optional"] is None

    def test_context_serializes_lists(self, console_config: LogConfig, capsys):
        """Test that context serializes lists."""
        logger = initialize(console_config)

        logger.info("test", tags=["python", "logging"])

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert entry["context"]["tags"] == ["python", "logging"]

    def test_context_serializes_nested_dicts(self, console_config: LogConfig, capsys):
        """Test that context serializes nested dictionaries."""
        logger = initialize(console_config)

        logger.info("test", metadata={"region": "us-east-1", "az": "us-east-1a"})

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert entry["context"]["metadata"]["region"] == "us-east-1"
            assert entry["context"]["metadata"]["az"] == "us-east-1a"


class TestExceptionFormatting:
    """Tests for exception field formatting."""

    def test_exception_has_type(self, console_config: LogConfig, capsys):
        """Test that exception includes type."""
        logger = initialize(console_config)

        try:
            raise ValueError("test error")
        except ValueError as e:
            logger.error("error occurred", exception=e)

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "exception" in entry
            assert "type" in entry["exception"]
            assert entry["exception"]["type"] == "ValueError"

    def test_exception_has_message(self, console_config: LogConfig, capsys):
        """Test that exception includes message."""
        logger = initialize(console_config)

        try:
            raise ValueError("test error message")
        except ValueError as e:
            logger.error("error occurred", exception=e)

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "exception" in entry
            assert "message" in entry["exception"]
            assert entry["exception"]["message"] == "test error message"

    def test_exception_has_stacktrace_field(self, console_config: LogConfig, capsys):
        """Test that exception includes stacktrace field."""
        logger = initialize(console_config)

        try:
            raise RuntimeError("test")
        except RuntimeError as e:
            logger.error("error", exception=e)

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "exception" in entry
            assert "stacktrace" in entry["exception"]
            # May be empty in current implementation
            assert isinstance(entry["exception"]["stacktrace"], str)


class TestDurationFormatting:
    """Tests for duration_ms formatting."""

    def test_duration_is_number(self, console_config: LogConfig, capsys):
        """Test that duration_ms is a number."""
        logger = initialize(console_config)

        with logger.timer("operation"):
            pass

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            assert "context" in entry
            if "duration_ms" in entry["context"]:
                assert isinstance(entry["context"]["duration_ms"], (int, float))
                assert entry["context"]["duration_ms"] >= 0

    def test_duration_has_precision(self, console_config: LogConfig, capsys):
        """Test that duration has sub-millisecond precision."""
        logger = initialize(console_config)

        import time

        with logger.timer("short operation"):
            time.sleep(0.001)  # 1ms

        captured = capsys.readouterr()
        if captured.out:
            entry = json.loads(captured.out.strip())

            if "duration_ms" in entry.get("context", {}):
                # Should have fractional milliseconds
                duration = entry["context"]["duration_ms"]
                assert isinstance(duration, float)


class TestFieldOrdering:
    """Tests for consistent field ordering (not required but nice)."""

    def test_required_fields_first(self, console_config: LogConfig, capsys):
        """Test that required fields appear early in output."""
        logger = initialize(console_config)

        logger.info("test")

        captured = capsys.readouterr()
        if captured.out:
            # JSON dict ordering may not be guaranteed in all Python versions
            # but we can check fields exist
            entry = json.loads(captured.out.strip())

            required_fields = [
                "timestamp",
                "level",
                "message",
                "service",
                "file",
                "line",
                "function",
            ]

            for field in required_fields:
                assert field in entry

File: cpp/tests/test_formatter.cpp
/**
 * @file test_formatter.cpp
 * @brief JSON formatter tests
 *
 * Tests cover:
 * - JSON output format validation
 * - Required fields present (timestamp, level, message, service, file, line, function)
 * - Context serialization
 * - Exception formatting
 * - Duration formatting
 */

#include <catch2/catch_test_macros.hpp>

#include <agora/log/logger.hpp>
#include <agora/log/config.hpp>

#include <filesystem>
#include <fstream>
#include <thread>
#include <chrono>
#include <nlohmann/json.hpp>

using namespace agora::log;
using json = nlohmann::json;
namespace fs = std::filesystem;

class FormatterTestFixture {
public:
    fs::path test_log_dir;
    fs::path test_log_file;

    void SetUp() {
        test_log_dir = fs::temp_directory_path() / "agora_formatter_tests";
        fs::create_directories(test_log_dir);
        test_log_file = test_log_dir / "formatter.log";
    }

    void TearDown() {
        if (fs::exists(test_log_dir)) {
            fs::remove_all(test_log_dir);
        }
        shutdown();
    }

    json read_first_entry() {
        flush();  // Flush handlers before reading (keeps logging active)
        std::ifstream file(test_log_file);
        std::string line;
        std::getline(file, line);
        return json::parse(line);
    }
};

TEST_CASE("JSON formatter - required fields", "[formatter][json]") {
    FormatterTestFixture fixture;
    fixture.SetUp();

    auto config = Config{};
    config.service_name = "test-service";
    config.environment = "test-env";
    config.version = "1.2.3";
    config.file_path = fixture.test_log_file;
    config.console_enabled = false;

    auto result = initialize(config);
    REQUIRE(result.has_value());

    auto logger = get_logger("test.formatter");
    logger.info("Test message");

    shutdown();

    auto entry = fixture.read_first_entry();

    // REQUIRED fields
    REQUIRE(entry.contains("timestamp"));
    REQUIRE(entry.contains("level"));
    REQUIRE(entry.contains("message"));
    REQUIRE(entry.contains("service"));
    REQUIRE(entry.contains("environment"));
    REQUIRE(entry.contains("version"));
    REQUIRE(entry.contains("file"));      // REQUIRED
    REQUIRE(entry.contains("line"));      // REQUIRED
    REQUIRE(entry.contains("function"));  // REQUIRED
    REQUIRE(entry.contains("logger_name"));

    // Verify values
    REQUIRE(entry["level"] == "INFO");
    REQUIRE(entry["message"] == "Test message");
    REQUIRE(entry["service"] == "test-service");
    REQUIRE(entry["environment"] == "test-env");
    REQUIRE(entry["version"] == "1.2.3");

    fixture.TearDown();
}

TEST_CASE("JSON formatter - timestamp format", "[formatter][timestamp]") {
    FormatterTestFixture fixture;
    fixture.SetUp();

    auto config = Config{};
    config.service_name = "test";
    config.file_path = fixture.test_log_file;
    config.console_enabled = false;

    auto result = initialize(config);
    REQUIRE(result.has_value());

    auto logger = get_logger("test");
    logger.info("Test");

    shutdown();

    auto entry = fixture.read_first_entry();

    // Timestamp should be ISO 8601 format with microseconds
    std::string timestamp = entry["timestamp"];

    // Basic format check: YYYY-MM-DDTHH:MM:SS.ssssssZ
    REQUIRE(timestamp.length() >= 20);
    REQUIRE(timestamp.find('T') != std::string::npos);
    REQUIRE(timestamp.find('Z') != std::string::npos);

    fixture.TearDown();
}

TEST_CASE("JSON formatter - context serialization", "[formatter][context]") {
    FormatterTestFixture fixture;
    fixture.SetUp();

    auto config = Config{};
    config.service_name = "test";
    config.file_path = fixture.test_log_file;
    config.console_enabled = false;

    auto result = initialize(config);
    REQUIRE(result.has_value());

    auto logger = get_logger("test");

    logger.info("Test with context", {
        {"string_val", "hello"},
        {"int_val", std::int64_t{42}},
        {"double_val", 3.14},
        {"bool_val", true}
    });

    shutdown();

    auto entry = fixture.read_first_entry();

    REQUIRE(entry.contains("context"));
    auto context = entry["context"];

    REQUIRE(context["string_val"] == "hello");
    REQUIRE(context["int_val"] == 42);
    REQUIRE(context["double_val"] == 3.14);
    REQUIRE(context["bool_val"] == true);

    fixture.TearDown();
}

TEST_CASE("JSON formatter - exception formatting", "[formatter][exception]") {
    FormatterTestFixture fixture;
    fixture.SetUp();

    auto config = Config{};
    config.service_name = "test";
    config.file_path = fixture.test_log_file;
    config.console_enabled = false;

    auto result = initialize(config);
    REQUIRE(result.has_value());

    auto logger = get_logger("test");

    try {
        throw std::logic_error("Test exception");
    } catch (const std::exception& ex) {
        logger.error("Error occurred", ex);
    }

    shutdown();

    auto entry = fixture.read_first_entry();

    REQUIRE(entry.contains("exception"));
    auto exception = entry["exception"];

    REQUIRE(exception.contains("type"));
    REQUIRE(exception.contains("message"));

    REQUIRE(exception["type"] == "std::logic_error");
    REQUIRE(exception["message"] == "Test exception");

    fixture.TearDown();
}

TEST_CASE("JSON formatter - duration formatting", "[formatter][duration]") {
    FormatterTestFixture fixture;
    fixture.SetUp();

    auto config = Config{};
    config.service_name = "test";
    config.file_path = fixture.test_log_file;
    config.console_enabled = false;

    auto result = initialize(config);
    REQUIRE(result.has_value());

    auto logger = get_logger("test");

    {
        auto timer = logger.timer("Test operation");
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    shutdown();

    auto entry = fixture.read_first_entry();

    REQUIRE(entry.contains("duration_ms"));
    double duration = entry["duration_ms"];

    // Should be at least 10ms
    REQUIRE(duration >= 10.0);
    REQUIRE(duration < 100.0);

    fixture.TearDown();
}

TEST_CASE("JSON formatter - special characters escaping", "[formatter][escaping]") {
    FormatterTestFixture fixture;
    fixture.SetUp();

    auto config = Config{};
    config.service_name = "test";
    config.file_path = fixture.test_log_file;
    config.console_enabled = false;

    auto result = initialize(config);
    REQUIRE(result.has_value());

    auto logger = get_logger("test");

    logger.info("Message with \"quotes\" and \n newlines", {
        {"key_with_\"quotes\"", "value"},
        {"backslash", "C:\\Path\\To\\File"}
    });

    shutdown();

    // Should be valid JSON
    REQUIRE_NOTHROW(fixture.read_first_entry());

    fixture.TearDown();
}
